File: Source/Remote Control Server/Remote.vb
Imports System.Management
Imports System.Runtime.InteropServices

Module Remote

    Public Declare Function SetCursorPos Lib "user32" (ByVal X As Integer, ByVal Y As Integer) As Integer
    Public Declare Sub mouse_event Lib "user32" Alias "mouse_event" (ByVal dwFlags As Integer, ByVal dx As Integer, ByVal dy As Integer, ByVal cButtons As Integer, ByVal dwExtraInfo As Integer)
    Public Declare Unicode Sub keybd_event Lib "user32.dll" (ByVal bVk As Byte, ByVal bScan As Byte, ByVal dwFlags As Int32, ByVal dwExtraInfo As Int32)

    Public latestApi As Byte = 3
    Public lastCommand As Command = New Command

    Public Sub processCommand(ByRef command As Command)
        lastCommand = command

        Select Case command.api
            Case 3
                ApiV3.parseCommand(command)
            Case 2
                ApiV2.parseCommand(command)
            Case 1
                ApiV1.parseCommand(command)
            Case Else
                'Use latest available API
                ApiV3.parseCommand(command)
        End Select

        Server.status = "Active"
        Server.getApp(command.source).status = "Active"
        Server.getApp(command.source).isConnected = True
        Server.getApp(command.source).isActive = False
    End Sub

End Module

File: Source/MacOS/RemoteControlServer/remote.cpp
#include "apiv1.h"
#include "apiv2.h"
#include "apiv3.h"
#include "remote.h"
#include "server.h"

#include <QDebug>

Remote* Remote::instance = NULL;

Remote* Remote::Instance()
{
    if (!instance)
    {
        instance = new Remote();
    }
    return instance;
}

Remote::Remote()
{
    latestApi = 3;
    lastCommandIsInitialized = false;
}

void Remote::initializeLastCommand()
{
    lastCommand = new Command();
    lastCommandIsInitialized = true;
}

void Remote::processCommand(Command &command)
{
    lastCommand = &command;
    switch (command.api)
    {
    case 3:
        ApiV3::Instance()->parseCommand(command);
        break;
    case 2:
        ApiV2::Instance()->parseCommand(command);
        break;
    case 1:
        ApiV1::Instance()->parseCommand(command);
        break;
    default:
        ApiV3::Instance()->parseCommand(command);
        break;
    }
    Server::Instance()->status = "active";
    Server::Instance()->getApp(command.source)->status = "Active";
    Server::Instance()->getApp(command.source)->isConnected = false;
}
